### PR TITLE
Godot 3.0 primitives as resources for use with MeshInstance

### PR DESCRIPTION
**Pull request by BastiaanOlij**
I've ported PR #8578 to Godot 3 and as suggested by @reduz changed the implementation to resources subclassed from Mesh instead of Nodes.

There are a few small fixes in this, I had calculated the normals of the cylinders wrong and you now have more freedom on the prism to move the top of the triangle in any position between left (0.0) and right (1.0).

Currently my build of Godot 3 crashes on creating materials so I have no been able to test if there are any issues with the texturing and I have no updated my test project to Godot 3.0 yet as a result.

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
I will also be adding more objects in a later build, took me all morning to port this and I am running out of battery :)

One thing I do not like, and @reduz I hope you can make some suggestions here on how I can improve this, is that I rebuild the mesh on every property change. Is there a way to postpone this until the mesh is needed for the first time since the last property change?

**Comment by RameshRavone** [comment]
STATIC_CHECKS: Clang-format is not happy :smile:

**Comment by BastiaanOlij** [comment]
Yeah Clang and I still need to learn to get along :) I did manage to get it installed the other day so I need to start playing with it and get it to go over my sourcecode before I commit it :)

**Comment by nunodonato** [comment]
@BastiaanOlij check https://travis-ci.org/godotengine/godot/jobs/229333800 it tells you what you need to change to conform to clang style

**Comment by BastiaanOlij** [comment]
@nunodonato I actually got clang-format working (I think) so my next commit should *fingers crossed* have all those issues fixed :)
I tried doing it by hand for one of my other PRs and its annoying:)

**Comment by BastiaanOlij** [comment]
Ok, hopefully clang will love me now. I've also added a 3D plane which you can set segments for, together with a vertex shader that would make an ideal candidate for rendering height maps. I've also added a capsule object. Haven't been able to test if the textures are ok yet, recompiling on latest master atm, hoping it'll fix the crashing issue...

File: modules/primitives/register_types.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#include "capsule3d.h"
+#include "cube3d.h"
+#include "cylinder3d.h"
+#include "plane3d.h"
+#include "prism3d.h"
+#include "sphere3d.h"
+
+void register_primitives_types() {
+
+  ClassDB::register_class<Capsule3D>();
```
Comment (akien-mga) [inline modules/primitives/register_types.cpp]: Hm, you don't want clang-format to like you, do you? :p

**Comment by akien-mga** [comment]
Still got a couple things to fix to make clang-format happy: https://travis-ci.org/godotengine/godot/jobs/229399863
You should really look into that precommit hook to let it do formatting automatically :)

**Comment by BastiaanOlij** [comment]
@akien-mga , how weird, I thought I had ran it on there, I must have made some last minute changes that caused something to misalign. 

Yeah indeed, I need to figure out the precommit thingy, I was happy enough to get clang-format to get installed :)

**Comment by BastiaanOlij** [comment]
Yeah! all green ticks:)

**Comment by BastiaanOlij** [comment]
@reduz I had a quick look this morning about changing this over on the new structure as you suggested in #8578 , any suggestions of a class to look at that gives a bit of an idea of how the new approach works? Looks like you retired quadmesh? :)

I'll keep puzzling away in the mean while.

**Comment by reduz** [comment]
I can rewrite quadmesh for reference

**Comment by BastiaanOlij** [comment]
@reduz only if its not too much trouble else I'm sure we'll catch eachother on IRC. Just need to get an idea of what the new structure is like.

**Comment by BastiaanOlij** [comment]
BROKEN -- Nearly all removed, will be rectified shortly

Just commited first attempt at changing implementation over to new approach. just working on cube3d atm

File: scene/resources/cubemesh.cpp [hunk 2]
```diff
@@ -0,0 +1,396 @@
+/*************************************************************************/
+/*  cube3d.cpp                                                           */
```
Comment (bojidar-bg) [inline scene/resources/cubemesh.cpp]: Yet another wrong filename in the header :wink:

Comment (BastiaanOlij) [inline scene/resources/cubemesh.cpp]: Lolz, yeah after discussing with reduz we felt renaming this to cubemesh made more sense. The originals were all called something3d

But I'm far more interested in the golden tip that allows me to figure out what I'm doing wrong and why I'm not actually getting a cube created! :)

File: scene/resources/cubemesh.cpp [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+	Array arr;
+	arr.resize(VS::ARRAY_MAX);
+	arr[VS::ARRAY_VERTEX] = points;
+	arr[VS::ARRAY_NORMAL] = normals;
+	arr[VS::ARRAY_TANGENT] = tangents;
+	arr[VS::ARRAY_TEX_UV] = uvs;
+	arr[VS::ARRAY_INDEX] = indices;
+
+	// out with the old
+	while (VisualServer::get_singleton()->mesh_get_surface_count(mesh) > 0) {
```
Comment (reduz) [inline scene/resources/cubemesh.cpp]: Any request of info from VisualServer (in this case mesh_get_surface_count()) stalls the rendering thread, so it's better avoided

**Comment by BastiaanOlij** [comment]
Ok, we have a working cubemesh!! :)

I'll start working on converting the other primitives later in the week, should be pretty smooth sailing from here on. One of the suggestions made on IRC was to add a class in between Mesh and these primitives shapes as there will be some overlapping code. Not sure what a good name would be, suggestions are welcome :)

**Comment by Zireael07** [comment]
PrimitiveMesh sound good as a class name?

**Comment by bojidar-bg** [comment]
I guess GeneratedMesh is better?

**Comment by BastiaanOlij** [comment]
The thing I'm most worried about regarding the name is that it has the potential to put people on the wrong foot. The thing that would make these classes easy to implement with the largest amount of common code in its base class would be to make the assumption any subclass will always have a single surface and single material. The subclass then simply implements the code that generates this surface.

GeneratedMesh suggests it would be able to do more. If I were to implement a dual contouring algorithm to generate the mesh for a terrain section (something on my wishlist of things to do) I might wish to create multiple surfaces. 

The opposite goes for PrimitiveMesh, the base class has the potential to be used for more then just generating primitive shapes.

I'm probably overthinking it ;)

**Comment by BastiaanOlij** [comment]
4 out of the original 6 mesh types have now been converted to the new approach. 

@reduz asked to also add the original QuadMesh to this. PlaneMesh possibly replaces this as it can create a Quad if you subdivide each side by 1.

**Comment by bojidar-bg** [comment]
Would the Quad and the TestCube nodes be obsoleted by the change?

**Comment by reduz** [comment]
Yes they should be removed

**Comment by BastiaanOlij** [comment]
I think it's all done. Note that there is a test project to test this with here:

https://github.com/BastiaanOlij/TestPrimitives

**Comment by rraallvv** [comment]
Does anyone know of a demo video showing off this features? Thanks.

**Comment by Zireael07** [comment]
@rraallvv: you don't really need a demo video. Create a mesh instance node, click the arrow next to mesh: and voila, you can see all the primitives available.

**Comment by rraallvv** [comment]
@Zireael07 I see... maybe it isn't available for 2.1, I'll try 3.0 then, thanks.

**Comment by Zireael07** [comment]
This is only for 3.0, yes.

**Comment by aaronfranke** [comment]
The Capsule primitive (and collider) are the wrong orientation by default. https://github.com/godotengine/godot/issues/5813

**Pull request merged**